src/main/java/htsjdk/samtools/util/nio/DeleteOnExitPathHook.java

```diff
@@ -0,0 +1,63 @@
+package htsjdk.samtools.util.nio;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+
+/**
+ * Class to hold a set of {@link Path} to be delete on the JVM exit through a shutdown hook.
+ *
+ * <p>This class is a modification of {@link java.io.DeleteOnExitHook} to handle {@link Path}
+ * instead of {@link java.io.File}.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+public class DeleteOnExitPathHook {
+    private static LinkedHashSet<Path> paths = new LinkedHashSet<>();
+    static {
+        Runtime.getRuntime().addShutdownHook(new Thread(DeleteOnExitPathHook::runHooks));
+    }
+
+    private DeleteOnExitPathHook() {}
+
+    /**
+     * Adds a {@link Path} for deletion on JVM exit.
+     *
+     * @param path path to be deleted.
+     *
+     * @throws IllegalStateException if the shutdown hook is in progress.
+     */
+    public static synchronized void add(Path path) {
+        if(paths == null) {
+            // DeleteOnExitHook is running. Too late to add a file
+            throw new IllegalStateException("Shutdown in progress");
+        }
+
+        paths.add(path);
+    }
+
+    static void runHooks() {
+        LinkedHashSet<Path> thePaths;
+
+        synchronized (DeleteOnExitPathHook.class) {
+            thePaths = paths;
+            paths = null;
+        }
+
+        ArrayList<Path> toBeDeleted = new ArrayList<>(thePaths);
+
+        // reverse the list to maintain previous jdk deletion order.
+        // Last in first deleted.
+        Collections.reverse(toBeDeleted);
+        for (Path path : toBeDeleted) {
+            try {
+                Files.delete(path);
+            } catch (IOException | RuntimeException e) {
+                // do nothing if cannot be deleted, because it is a shutdown hook
```

Comment:
It seems wrong to throw away an exception here. Can it safely be logged?

Reply:
This is what the File version of the hook is doing (because the result of the `File.delete` is ignored there). In addition, because it is a shutdown-hook, the JVM will crash on exit if we throw them (see an example in https://github.com/samtools/htsjdk/issues/1029).

I think that we can re-write this block as:

```java
...
try {
    Files.delete(path);
} catch (IOException | RuntimeException e) {
    System.err.println(String.format("Error deleting on exit %s: %s",
                path.toUri().toString(),
                e.getMessage())
    );
    e.printStackTrace();
}
...
```

But I think that this is too verbose, and actually it will be a difference with the `File` version.

Reply:
This isn't what the `File` version is doing; your code catches `RuntimeException` and the `File` code does not. This is a pretty big difference in behavior.

Maybe logging a stack trace isn't necessary here, though. Also we don't plan to delete GCS bucket entries on exit, correct? If so you can format the output using `String.format(".... %s ...", path);` and it will look fine.

Reply:
I am catching `RuntimeException` because of the `SecurityException` thrown by `Files.delete`.

Next commit I will catch only that one, without the logging. I think that it is too verbose for just saying that a temp file was not removed if we add them (e.g., if it does not exists or the directory isn't empty).

Reply:
> I am catching RuntimeException because of the SecurityException thrown by Files.delete.

OK, so this is a required difference between our delete-on-exit hook and the one in `java.io`?

> I think that it is too verbose for just saying that a temp file was not removed if we add them (e.g., if it does not exists or the directory isn't empty).

That is not how `File.delete()` works. It will not throw an exception if the file doesn't exist or if the directory isn't empty, it will only return `false` to indicate failure.

Reply:
But `File.deleteOnExit` uses a hook that ignores the value of `File.delete`. This hook should be use for mimic deleteOnExit for Path, that's why is ignoring the result of delete...